### PR TITLE
Add input output aliasing to CustomCallOp

### LINKED ISSUE #235: Consider adding input output aliasing to CustomCallOp
Corresponding MHLO change: https://github.com/tensorflow/mlir-hlo/commit/6741953eb7c0fe52a99c5d85253e31524ea058c3.

### PR DESCRIPTION
**Pull request by subhankarshah**
Closes #235

### CODE REVIEW

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/dialect/StablehloAttrs.td [hunk 1]
```diff
@@ -79,6 +79,50 @@ def StableHLO_ConvDimensionNumbers : AttrDef<StableHLO_Dialect, "ConvDimensionNu
...
+  let parameters = (ins
+    StableHLO_Dim:$outputTupleIndices,
+    "int64_t":$operandIndex,
+    StableHLO_Dim:$operandTupleIndices
+  );
+  let assemblyFormat = "`<` "
+                       "`output_tuple_indices` `=` `[` $outputTupleIndices `]` `,`"
+                       "`operand_index` `=` $operandIndex `,`"
+                       "`operand_tuple_indices` `=` `[` $operandTupleIndices `]`"
+                       "`>`";
```
[the remark this answers is not in the document]
Comment (subhankarshah) [reply stablehlo/dialect/StablehloAttrs.td:123]: this was updated recently, fixed.

[the remark this answers is not in the document]
Comment (subhankarshah) [reply stablehlo/dialect/StablehloAttrs.td]: `def StableHLO_Dim : ArrayRefParameter<"int64_t", "Dimension">;`

does not have custom parser yet

[the remark this answers is not in the document]
Comment (subhankarshah) [reply stablehlo/dialect/StablehloAttrs.td]: @burmako

[the remark this answers is not in the document]
Comment (GleasonK) [reply stablehlo/dialect/StablehloAttrs.td]: Good catch. I created #485 for this feedback. This function should be moved from MHLO to the shared AssemblyFormat.

[the remark this answers is not in the document]
Comment (burmako) [reply stablehlo/dialect/StablehloAttrs.td]: Thank you for looking into this! SGTM to use the old assembly format in the meanwhile.

File: stablehlo/integrations/python/tests/stablehlo.py [hunk 2]
```diff
@@ -35,7 +35,6 @@ def test_channel_handle():
   assert attr.handle == 1
   assert attr.channel_type == 2
 
-
```
Comment (burmako) [inline stablehlo/integrations/python/tests/stablehlo.py:38]: You probably didn't mean to remove this line?

File: stablehlo/dialect/StablehloOps.td [hunk 3]
```diff
@@ -2021,6 +2026,19 @@ def StableHLO_CustomCallOp: StableHLO_Op<"custom_call",
...
+  let builders = [
+    OpBuilder<(ins
+      "::mlir::TypeRange":$result_type, "::mlir::ValueRange":$operands,
+      "::mlir::StringAttr":$call_target_name,
+      "::mlir::BoolAttr":$has_side_effect,
+      "::mlir::StringAttr":$backend_config,
+      "::mlir::stablehlo::CustomCallApiVersionAttr":$api_version,
+      "::mlir::ArrayAttr":$called_computations,
+      "::mlir::ArrayAttr":$operand_layouts,
+      "::mlir::ArrayAttr":$result_layouts)>];
```
[the remark this answers is not in the document]
Comment (burmako) [reply stablehlo/dialect/StablehloOps.td:2071]: This comment doesn't seem to have been addressed.

**Review by burmako — CHANGES_REQUESTED** [review]

File: stablehlo/dialect/StablehloOps.td (hunk 3 above)
[the remark this answers is not in the document]
Comment (subhankarshah) [reply stablehlo/dialect/StablehloOps.td]: I somehow did no stage this change, apologies.

**Review by burmako — APPROVED** [review]

**Comment by burmako** [comment]
Rationale for approval is the same as in #272 and #388: this is a non-controversial backward-compatible change, and accepting it doesn't violate any of the existing commitments (it sticks to existing HLO semantics, and it is compatible with the extent of the current compatibility commitments).

**Pull request merged**